docs/book/src/clusterctl/provider-contract.md

```diff
@@ -150,8 +150,33 @@ will look for objects to reconcile.
 
 #### Variables
 
-The components YAML can contain environment variables matching the regexp `\${\s*([A-Z0-9_]+)\s*}`; it is highly
-recommended to prefix the variable name with the provider name e.g. `${ AWS_CREDENTIALS }`
+The components YAML can contain environment variables matching the format ${VAR}; it is highly
+recommended to prefix the variable name with the provider name e.g. `${AWS_CREDENTIALS}`
+
+<aside class="note warning">
+
+<h1>Warning</h1>
+
+`clusterctl` previously supported variables that matched the following
+formats: `${ VAR }`, `${ VAR}`,`${VAR }` and `${VAR$FOO}`.
+
+The first three formats are currently supported but will be deprecated in the
+near future.
+<br/>
+The last format is no longer supported.
+</aside>
+
+`clusterctl` uses the library [drone/envsubst][drone-envsubst] to perform
+variable substitution. In order to provide defaults to the variables, one of
+the following conventions may be used.
+
+```
+${VAR:=default}
+${VAR=default}
+${VAR:-default}
```

Comment:
I don't know the specific of the library we are using, but AFAIK as i know there are differences betwenn `:=`, `=`, `:-`
I have found some documentation here https://github.com/a8m/envsubst#docsm but I think that we should provide something similar for every transformation supported by drone.

Reply:
I wanted to document just the "default" transformations here since that's most relevant for our use cases and if users were interested in other transformations then they could reference the original drone/envsubst docs. Another reason I was avoiding document every transformation was just in case drone/envsubst changed over time for whatever reason.

I'll definitely test out the other formats `=`, `:-` and document those better. Thanks!

Reply:
I can provide this link as a source of documentation for the other transformations instead of re-documenting it here as well.

Reply:
On second thought, it seems that the drone/envsubst implementation does not follow the spec here: http://www.tldp.org/LDP/abs/html/refcards.html#AEN22728

So I'll just update the docs as much as I can for now.

Reply:
The current code treats all the above as the same syntax regardless if the `VAR` is set or empty.

https://github.com/drone/envsubst/blob/a79604a7a2149df6ff9f6b20b2ff188112a777b6/template.go#L150-L151
```golang
	case "=", ":=", ":-":
		return toDefault
```

https://github.com/drone/envsubst/blob/a79604a7a2149df6ff9f6b20b2ff188112a777b6/funcs.go#L54-L59
```golang
func toDefault(s string, args ...string) string {
	if len(s) == 0 && len(args) == 1 {
		s = args[0]
	}
	return s
}
```

I'll see if I can submit PRs to drone/envsubst to fix them.

Reply:
kk, for getting started let's document the status quo: what are the syntax supported by drone and all the nuances (different syntax doing the same)
Then, if there is a need/a use case to be addressed, we can start opening issues/sending PRs to drone